broker-core/src/main/java/io/zeebe/broker/workflow/processor/CatchEventOutput.java

```diff
@@ -0,0 +1,199 @@
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package io.zeebe.broker.workflow.processor;
+
+import static io.zeebe.util.buffer.BufferUtil.bufferAsString;
+import static io.zeebe.util.buffer.BufferUtil.cloneBuffer;
+
+import io.zeebe.broker.incident.data.ErrorType;
+import io.zeebe.broker.logstreams.processor.TypedBatchWriter;
+import io.zeebe.broker.subscription.command.SubscriptionCommandSender;
+import io.zeebe.broker.workflow.data.TimerRecord;
+import io.zeebe.broker.workflow.model.element.ExecutableIntermediateCatchElement;
+import io.zeebe.broker.workflow.model.element.ExecutableMessage;
+import io.zeebe.broker.workflow.state.ElementInstance;
+import io.zeebe.broker.workflow.state.TimerInstance;
+import io.zeebe.broker.workflow.state.WorkflowState;
+import io.zeebe.broker.workflow.state.WorkflowSubscription;
+import io.zeebe.msgpack.query.MsgPackQueryProcessor;
+import io.zeebe.msgpack.query.MsgPackQueryProcessor.QueryResult;
+import io.zeebe.msgpack.query.MsgPackQueryProcessor.QueryResults;
+import io.zeebe.protocol.intent.TimerIntent;
+import io.zeebe.util.sched.clock.ActorClock;
+import java.time.Duration;
+import java.util.List;
+import org.agrona.DirectBuffer;
+
+/**
+ * FIXME: find a better name
+ *
+ * <p>Split into multiple files once we have a reasonable amount of event triggers.
+ */
+public class CatchEventOutput {
+  private final WorkflowState state;
+  private final SubscriptionCommandSender subscriptionCommandSender;
+
+  public CatchEventOutput(
+      WorkflowState state, SubscriptionCommandSender subscriptionCommandSender) {
+    this.state = state;
+    this.subscriptionCommandSender = subscriptionCommandSender;
+  }
+
+  public void unsubscribeFromCatchEvents(
+      BpmnStepContext<?> context, final List<? extends ExecutableIntermediateCatchElement> events) {
+    if (events.size() > 1) {
+      context.getOutput().useBatchWriter();
+    }
+
+    // at the moment, the way the state is handled we don't need specific event information to
+    // unsubscribe from an event trigger, but once messages are supported it will be necessary.
+    unsubscribeFromTimerEvents(
+        context.getElementInstance().getKey(), context.getOutput().getBatchWriter());
+    unsubscribeFromMessageEvent(context);
+  }
+
+  public void subscribeToCatchEvents(
+      BpmnStepContext<?> context, final List<? extends ExecutableIntermediateCatchElement> events) {
+    if (events.size() > 1) {
+      context.getOutput().useBatchWriter();
```

Comment:
This requires that the calling code did not yet stage any other event, which seems very fragile. Is there a better solution?

Reply:
Let the calling code be responsible for this, I suppose. The main reason why this is here is because we have differing APIs for batch writers/single writers...

Simplest fix is to check here again if we use batch writer or not and then use the correct API, as we do with the `EventOutput`, but that's not ideal either.

I can reduce how fragile this is by delegating responsibility to the calling code, and adding an assert here to check if we're using the batch writer or not.